sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +385,464 @@ def transform(self, X):
+                            estimator=None,
+                            min_std=1e-6):
+        """Imputes a single feature from the others provided.
+
+        This function predicts the missing values of one of the features using
+        the current estimates of all the other features. The `estimator` must
+        support `return_std=True` in its ``predict`` method for this function
+        to work.
+
+        Parameters
+        ----------
+        X_filled : array-like
+            Input data with the most recent imputations.
+
+        mask_missing_values : array-like
+            Input data's missing indicator matrix.
+
+        this_feat_ind : integer
+            Index of the feature currently being imputed.
+
+        other_feat_inds : array-like
+            Indices of the features to be used in imputing `this_feat_ind`
+
+        estimator : estimator with sklearn API, optional (default=None)
+            The estimator to use at this step of the round-robin imputation.
+            As of now, it must support `return_std` in its ``predict`` method.
+            If None, it will be fit, otherwise not.
+
+        min_std : float, optional (default=1e-6)
+            The smallest allowable standard deviation for the posterior
+            sampling step.
+
+        Returns
+        -------
+        X_filled : array-like
+            Input data with `X_filled[missing_row_mask, this_feat_ind]`
+            updated.
+
+        estimator : estimator with sklearn API
+            The fitted estimator used to impute
+            `X_filled[missing_row_mask, this_feat_ind]`.
+        """
+
+        # if nothing is missing, just return the default
+        rng = check_random_state(self.random_state)
+        if mask_missing_values[:, this_feat_ind].sum() == 0:
+            return X_filled, estimator
+        missing_row_mask = mask_missing_values[:, this_feat_ind]
+
+        # if no estimator provided, instantiate a new one and fit
+        if estimator is None:
+            X_train = X_filled[:, other_feat_inds][~missing_row_mask]
+            y_train = X_filled[:, this_feat_ind][~missing_row_mask]
+            if np.std(y_train) > 0:
+                estimator = clone(self.estimator_)
+                estimator.fit(X_train, y_train)
+            else:
+                estimator = DummyRegressor()
+                estimator.fit(X_train, y_train)
+
+        # get posterior samples
+        X_test = X_filled[:, other_feat_inds][missing_row_mask]
+        mus, sigmas = estimator.predict(X_test, return_std=True)
+        sigmas = np.maximum(sigmas, min_std)
+        imputed_values = rng.normal(mus, sigmas)
+
+        # clip the values (np.clip ignores np.nans)
+        imputed_values = np.clip(imputed_values,
+                                 self.min_value_,
+                                 self.max_value_)
+
+        # update the feature
+        X_filled[missing_row_mask, this_feat_ind] = imputed_values
+        return X_filled, estimator
```

Comment:
This is sort of tricky to do because which estimator gets used depends on the current state we are in: 
```
            if np.std(y_train) > 0:
                estimator = clone(self.estimator_)
                estimator.fit(X_train, y_train)
            else:
                estimator = DummyRegressor()
                estimator.fit(X_train, y_train)
```
and `y_train` gets defined inside this function. What's the issue with the estimator being returned?